### PR TITLE
Shutdown hook

### LINKED ISSUE #22: Listen to ApplicationLifecycle addStopHook
As [this](https://github.com/akka/akka/issues/16495) issue states, Akka can invoke scheduled task even though application has initiated STOP sequence. If one is using EntityManager, this causes nasty exception:
```
java.lang.IllegalStateException: EntityManagerFactory is closed
        at org.hibernate.jpa.internal.EntityManagerFactoryImpl.validateNotClosed(EntityManagerFactoryImpl.java:388)
        at org.hibernate.jpa.internal.EntityManagerFactoryImpl.internalCreateEntityManager(EntityManagerFactoryImpl.java:342)
        at org.hibernate.jpa.internal.EntityManagerFactoryImpl.createEntityManager(EntityManagerFactoryImpl.java:313)
        at play.db.jpa.DefaultJPAApi.em(DefaultJPAApi.java:71)
        at play.db.jpa.DefaultJPAApi.withTransaction(DefaultJPAApi.java:123)
        at play.db.jpa.DefaultJPAApi.withTransaction(DefaultJPAApi.java:80)
        at play.db.jpa.JPA.withTransaction(JPA.java:127)
        at controllers.sitemaps.SitemapUrlProvider.addUrls(SitemapUrlProvider.java:58)
        at com.edulify.modules.sitemap.SitemapTask.run(SitemapTask.java:35)
        at akka.actor.LightArrayRevolverScheduler$$anon$2$$anon$1.run(Scheduler.scala:242)
        at akka.actor.LightArrayRevolverScheduler$TaskHolder.run(Scheduler.scala:476)
        at akka.actor.LightArrayRevolverScheduler$$anonfun$close$1.apply(Scheduler.scala:282)
        at akka.actor.LightArrayRevolverScheduler$$anonfun$close$1.apply(Scheduler.scala:281)
        at scala.collection.Iterator$class.foreach(Iterator.scala:742)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1194)
        at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at akka.actor.LightArrayRevolverScheduler.close(Scheduler.scala:280)
        at akka.actor.ActorSystemImpl.stopScheduler(ActorSystem.scala:689)
        at akka.actor.ActorSystemImpl$$anonfun$liftedTree2$1$1.apply$mcV$sp(ActorSystem.scala:618)
        at akka.actor.ActorSystemImpl$$anonfun$liftedTree2$1$1.apply(ActorSystem.scala:618)
        at akka.actor.ActorSystemImpl$$anonfun$liftedTree2$1$1.apply(ActorSystem.scala:618)
        at akka.actor.ActorSystemImpl$$anon$3.run(ActorSystem.scala:642)
        at akka.actor.ActorSystemImpl$TerminationCallbacks$$anonfun$run$1.runNext$1(ActorSystem.scala:809)
        at akka.actor.ActorSystemImpl$TerminationCallbacks$$anonfun$run$1.apply$mcV$sp(ActorSystem.scala:812)
        at akka.actor.ActorSystemImpl$TerminationCallbacks$$anonfun$run$1.apply(ActorSystem.scala:805)
        at akka.actor.ActorSystemImpl$TerminationCallbacks$$anonfun$run$1.apply(ActorSystem.scala:805)
        at akka.util.ReentrantGuard.withGuard(LockUtil.scala:15)
        at akka.actor.ActorSystemImpl$TerminationCallbacks.run(ActorSystem.scala:805)
        at akka.actor.ActorSystemImpl$$anonfun$terminationCallbacks$1.apply(ActorSystem.scala:639)
        at akka.actor.ActorSystemImpl$$anonfun$terminationCallbacks$1.apply(ActorSystem.scala:639)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
        at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:91)
        at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
        at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:91)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:90)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
What you need to do is to inject ```ApplicationLifecycle``` and add a stopHook: 
```
private boolean stopped = false;
@Inject
public MyClass(ApplicationLifecycle lifecycle){
	lifecycle.addStopHook(() -> {
	        stopped=true;
	        return F.Promise.pure(null);
	});
}

@Override
public void run() {
	if(stopped == false){
		execute();
	}
}
```

### PR DESCRIPTION
**Pull request by KarelCemus**
This PR resolves #22 and workaround the Akka bug.

Besides the bugfix, I bumped versions of several dependencies, refactored duplicated code and migrated deprecated code.

### CODE REVIEW

File: app/com/edulify/modules/sitemap/SitemapTask.java [hunk 1]
```diff
@@ -1,35 +1,39 @@
...
 import java.util.List;
+import java.util.concurrent.CompletableFuture;
 
 public class SitemapTask implements Runnable {
 
     private SitemapConfig sitemapConfig;
     private SitemapProviders sitemapProviders;
 
+    // Indicates the application is running, see #22 for more details
+    private boolean running = true;
```
Comment (marcospereira) [inline app/com/edulify/modules/sitemap/SitemapTask.java]: I think I prefer the name `shuttingDown` here. It is more clear about what this variable will be used to.

File: build.sbt [hunk 2]
```diff
@@ -1,8 +1,8 @@
 name := "sitemap-module"
 
-version := "2.1.0"
+version := "2.1.1-SNAPSHOT"
```
Comment (marcospereira) [inline build.sbt]: Since there are changes in public APIs, I prefer to have a `2.2.0` version.

File: project/plugins.sbt [hunk 3]
```diff
@@ -4,6 +4,6 @@ logLevel := Level.Warn
 resolvers += "iBiblio Maven" at "http://mirrors.ibiblio.org/maven2/"
 
 // Use the Play sbt plugin for Play projects
-addSbtPlugin("com.typesafe.play" % "sbt-plugin" % System.getProperty("play.version", "2.5.0"))
+addSbtPlugin("com.typesafe.play" % "sbt-plugin" % System.getProperty("play.version", "2.5.1"))
```
Comment (marcospereira) [inline project/plugins.sbt]: Play 2.5.2 was already released. :-)

**Comment by marcospereira** [comment]
Thank you so much, @KarelCemus. See my comments.

**Comment by KarelCemus** [comment]
I there anything I can do to have this merged?

**Comment by marcospereira** [comment]
Thank you, @KarelCemus!

**Pull request merged**